src/operator-run.c

```diff
@@ -2034,23 +2041,226 @@ void xnn_compute_f16_qd8_convert(
   context->convert_ukernel(n, input, output, &params);
 }
 
+#define KAI_ASSERT(x)                                                                                                  \
+    do {                                                                                                               \
+        if (!(x)) {                                                                                                    \
+            exit(EXIT_FAILURE);                                                                                        \
+        }                                                                                                              \
+    } while (0)
+
+#define KAI_UNUSED(x) (void)(x)
+#define KAI_MIN(a, b) (((a) < (b)) ? (a) : (b))
+#define KAI_MAX(a, b) (((a) > (b)) ? (a) : (b))
+
+// The following functions should not be here because this file is not
+// target specific.
+size_t kai_get_lhs_offset_lhs_quant_pack_qai8dxp_f32(
+  size_t m_idx,
+  size_t lhs_stride)
+{
+  return m_idx * lhs_stride;
+}
+
+size_t kai_get_lhs_packed_offset_lhs_quant_pack_qai8dxp_f32(
+  size_t m_idx,
+  size_t k,
+  size_t mr,
+  size_t kr,
+  size_t sr)
+{
+  KAI_ASSERT(k % kr == 0);
+  KAI_UNUSED(kr);
+
+  const size_t kai_num_bytes_per_multiplier = sizeof(float);
+  const size_t kai_num_bytes_per_offset = sizeof(int32_t);
+  const size_t dst_y = (m_idx / mr);
+
+  const size_t dst_stride = mr * (k * sizeof(int8_t) + kai_num_bytes_per_multiplier + kai_num_bytes_per_offset);
+
+  // It points always at the beginning of the row
+  return dst_y * dst_stride;
+}
+
+void kai_run_lhs_quant_pack_qai8dxp_f32(
+  size_t m,
+  size_t k,
+  size_t mr,
+  size_t kr,
+  size_t sr,
+  size_t m_idx_start,
+  const float* restrict lhs,
+  size_t lhs_stride,
+  void* restrict lhs_packed)
+{
+  KAI_ASSERT(k % kr == 0);
+  KAI_ASSERT((kr % sr) == 0);
+
+  const size_t kai_num_bytes_per_multiplier = sizeof(float);
+  const size_t kai_num_bytes_per_offset = sizeof(int32_t);
+
+  if (m == 0) {
+      return;
+  }
+
+  const size_t num_rows = m;
+  const size_t num_cols = k;
+
+  const float* src_ptr = lhs;
+
+  const size_t dst_stride = mr * (k * sizeof(int8_t) + kai_num_bytes_per_multiplier + kai_num_bytes_per_offset);
+  const size_t k_block_len = kr / sr;
+
+  for (size_t row_idx = 0; row_idx < num_rows; ++row_idx) {
+    float32x4_t vmax0 = vdupq_n_f32(-FLT_MAX);
+    float32x4_t vmin0 = vdupq_n_f32(FLT_MAX);
```

Comment:
usually faster to initialize with a single value duped from the first float.  see f32-rminmax